modules/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/builtin/stringlib/AbstractRegexFunction.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.nativeimpl.builtin.stringlib;
+
+import org.ballerinalang.model.values.BStruct;
+import org.ballerinalang.natives.AbstractNativeFunction;
+import org.ballerinalang.util.exceptions.BallerinaException;
+
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * Abstract native function which serves as base class for writing Regex related native functions.
+ */
+public abstract class AbstractRegexFunction extends AbstractNativeFunction {
+
+    /**
+     * Validate Regex struct in Ballerina. Validate whether the string is in compiled form if not
+     * compile it and store as native data for later reference.
+     *
+     * @param regexStruct BStruct which holds compiled pattern.
+     * @return Pattern compiled Pattern
+     */
+    protected Pattern validatePattern(BStruct regexStruct) {
+        String regex = regexStruct.getStringField(0);
+        try {
+            Pattern pattern = (Pattern) regexStruct.getNativeData(REGEXConstants.COMPILED_REGEX);
+            if (pattern == null) {
+                pattern = Pattern.compile(regex);
+                regexStruct.addNativeData(REGEXConstants.COMPILED_REGEX, pattern);
+            }
+            return pattern;
+        } catch (PatternSyntaxException e) {
```

Comment:
You can't throw an error here. It will be mapped to Ballerina runtime error. As our best practice, we should return an error from this native function. So we need to handle this thrown error.

Reply:
First we thought of treat this exception as non recoverable error. Anyway +1 to make this consistent across other APIs.